### ISSUE TITLE
ABCP Pass determination is not correct.

### ISSUE THREAD

**Issue by fennecfox38** [fennecfox38/ArmyFitnessTestRecorder#1]
In ABCP, Body Fat measurement may be conducted only when SM is No-Go on Height/Weight.

The followings need to be fixed.
- [x] Hide body fat measurement when the user passed on Height/Weight.
- [x] Set 'pass' on ABCP when the user passed on either Height/Weight or Body Fat.

-----------------------------------------------------------------------------------------------------------

_Line 126-142 in ABCPFragment.java_
```java
public void onSexChanged(RadioGroup radioGroup, int checkedId){
    ArrayList<Item> list = itemList.getValue();
    if(checkedId == R.id.rBtn_male){
        record.sex = Sex.Male;
        list.remove(Item.HIPS);
        list.get(Item.ABDOMEN_WAIST).title = getString(R.string.abdomen);
    }
    else{
        record.sex = Sex.Female;
        list.get(Item.ABDOMEN_WAIST).title = getString(R.string.waist);
        list.add(HIPS);
    } // Attach and Detach Hips item
    itemList.setValue(list);
    adapter.notifyDataSetChanged();
    record.invalidate(list);//record.validate();
    binding.invalidateAll();
}
```

Item 'Hips' is being attached/detached in 'onSexChanged' in ABCPFragment.
The task 'Hide body fat measurement when the user passed on Height/Weight' requires 'itemList' to be updated more dynamically.

_Line 40-46 in ABCPRecord.java_
```java
public void invalidate(ArrayList<Item> items){
    updateRecord(items);
    height_weight = Standard.ABCP.isHWPassed(sex.ordinal(),ageGroup.ordinal(),height,weight);
    bodyFatPercentage = (sex==Sex.Male ? Standard.ABCP.maleBodyFat(height,neck, abdomen_waist) : Standard.ABCP.femaleBodyFat(height,neck, abdomen_waist,hips));
    bodyFatPass = Standard.ABCP.isBodyFatPassed(sex.ordinal(), ageGroup.ordinal(), bodyFatPercentage);
    isPassed = height_weight && bodyFatPass;
}
```

It will be better if ABCPRecord.invalidate(ArrayList\<Item\> items) is notified when any information is changed, and after that, void invalidate(ArrayList\<Item\> items) determines go/no-go, and update 'itemList', .

-----------------------------------------------------------------------------------------------------------

The task 'Set pass on ABCP when the user passed on either Height/Weight or Body Fat.' can be fixed easily by a simple modification on

_Line 45 in ABCPRecord.java_
```java
isPassed = height_weight && bodyFatPass;
```

But 'Body Fat Pass' needs 3rd state 'N/A' for when the user does not need to take it.
It is expected to implement another enum, or modify 'Body Fat indicating color'

_Line 138 in fragment_abcp.xml_
```xml
android:textColor="@{fragment.getPassedColor(fragment.record.bodyFatPass)}"
```

_Line 106-107 in ABCPFragment.java_
```java
public int getPassedColor(boolean isPassed){ return getResources().getColor(isPassed ? R.color.passed: R.color.failed); }
public String getPassed(boolean isPassed){ return getString(isPassed ? R.string.pass: R.string.fail); }
```

-----------------------------------------------------------------------------------------------------------

**Comment by fennecfox38** [comment]
This issue has been partly fixed by  **commit 'ABCP Determination corrected'  5ab882b2b61b82712505c1a53ba8156898bfd907**

As explained in the above commit 5ab882b2b61b82712505c1a53ba8156898bfd907,
> All 5 items are static, but just _ItemRecyclerAdapter_ shows selectively.
> _ItemRecyclerAdapter.itemCount_ has been implemented.
> itemCount will be updated on ABCPFragment, by the return value of _ABCPRecord.invalidate (ArrayList<Item> items)_

_ABCPRecord.invalidate (ArrayList<Item> items)_ determines how many items _ItemRecyclerAdapter_ should show based on Go/No-Go, and Sex, and returns it.

But as explained in the above commit 5ab882b2b61b82712505c1a53ba8156898bfd907,
> Still bodyfat requires 3rd state 'N/A', and needs to be reflected on view.

It still shows colored Body Fat Percentage from the last record when it is changed from H/W fail to H/W pass, and there is no state that explains 'N/A' properly.

------------------------------------------------------------------------------------------------------
### First Solution : Implementation of BodyFat State (Go, No-Go, N/A)

------------------------------------------------------------------------------------------------------
### Second Solution : 'N/A' Determination in real-time

------------------------------------------------------------------------------------------------------

So finally,

- [ ] Body Fat Indicator reflects 'N/A' state properly when the user passes Height/Weight
- [ ] Body Fat 'N/A' state can be reflected on DB.

**Comment by fennecfox38** [comment]
https://github.com/fennecfox38/ArmyFitnessTestRecorder/issues/1#issuecomment-693340705
> Body Fat Indicator reflects 'N/A' state properly when the user passes Height/Weight
Body Fat 'N/A' state can be reflected on DB.

The above two tasks have been done by the following two commits:
commit 7d9e26447f0d9236404548f485613533baa9230b commit 49b840613b9990da700d9f12617dad0b9745a46c

Second Solution: 'N/A' Determination in real-time is used.
implemented String getBodyFat(float bodyFatPercent, boolean hw)

Body Fat 'N/A' state can be reflected on DB.
ABCPRecord, ABCPDBHandler have been modified.

All tasks related to this issue are completed.

**Issue closed.**
